<!DOCTYPE html><html lang="ZH" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0,viewport-fit=cover"><title>比较基因组 | VEXYXING</title><meta name="author" content="VERYXING565"><meta name="copyright" content="VERYXING565"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="比较基因组各个步骤可能有多个方法都能达到相同目的，作为自学的，使用哪个方法是一个完全随缘的事情。 1 系统发育分析1.1 下载数据下载分析物种的fna基因组文件和gff注释文件。 过程如下：  收集需下载的基因组登入号  组装下载链接  下载基因组和注释（需要安装ascp）   由于不考虑自己重新注释的方法，我就将NCBI中某一个分类单位的所有可用的基因组全部下载下来（包括物种重复的）：  若在A">
<meta property="og:type" content="article">
<meta property="og:title" content="比较基因组">
<meta property="og:url" content="http://example.com/2025/02/12/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%842025-2-15/index.html">
<meta property="og:site_name" content="VEXYXING">
<meta property="og:description" content="比较基因组各个步骤可能有多个方法都能达到相同目的，作为自学的，使用哪个方法是一个完全随缘的事情。 1 系统发育分析1.1 下载数据下载分析物种的fna基因组文件和gff注释文件。 过程如下：  收集需下载的基因组登入号  组装下载链接  下载基因组和注释（需要安装ascp）   由于不考虑自己重新注释的方法，我就将NCBI中某一个分类单位的所有可用的基因组全部下载下来（包括物种重复的）：  若在A">
<meta property="og:locale">
<meta property="og:image" content="http://example.com/img/touxiang.jpg">
<meta property="article:published_time" content="2025-02-12T06:16:46.000Z">
<meta property="article:modified_time" content="2025-03-19T11:53:23.119Z">
<meta property="article:author" content="VERYXING565">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://example.com/img/touxiang.jpg"><script type="application/ld+json">{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "比较基因组",
  "url": "http://example.com/2025/02/12/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%842025-2-15/",
  "image": "http://example.com/img/touxiang.jpg",
  "datePublished": "2025-02-12T06:16:46.000Z",
  "dateModified": "2025-03-19T11:53:23.119Z",
  "author": [
    {
      "@type": "Person",
      "name": "VERYXING565",
      "url": "http://example.com/"
    }
  ]
}</script><link rel="shortcut icon" href="../../../../img/cat.jpg"><link rel="canonical" href="http://example.com/2025/02/12/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%842025-2-15/index.html"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="../../../../css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/node-snackbar/dist/snackbar.min.css" media="print" onload="this.media='all'"><script>
    (() => {
      
    const saveToLocal = {
      set: (key, value, ttl) => {
        if (!ttl) return
        const expiry = Date.now() + ttl * 86400000
        localStorage.setItem(key, JSON.stringify({ value, expiry }))
      },
      get: key => {
        const itemStr = localStorage.getItem(key)
        if (!itemStr) return undefined
        const { value, expiry } = JSON.parse(itemStr)
        if (Date.now() > expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return value
      }
    }

    window.btf = {
      saveToLocal,
      getScript: (url, attr = {}) => new Promise((resolve, reject) => {
        const script = document.createElement('script')
        script.src = url
        script.async = true
        Object.entries(attr).forEach(([key, val]) => script.setAttribute(key, val))
        script.onload = script.onreadystatechange = () => {
          if (!script.readyState || /loaded|complete/.test(script.readyState)) resolve()
        }
        script.onerror = reject
        document.head.appendChild(script)
      }),
      getCSS: (url, id) => new Promise((resolve, reject) => {
        const link = document.createElement('link')
        link.rel = 'stylesheet'
        link.href = url
        if (id) link.id = id
        link.onload = link.onreadystatechange = () => {
          if (!link.readyState || /loaded|complete/.test(link.readyState)) resolve()
        }
        link.onerror = reject
        document.head.appendChild(link)
      }),
      addGlobalFn: (key, fn, name = false, parent = window) => {
        if (!false && key.startsWith('pjax')) return
        const globalFn = parent.globalFn || {}
        globalFn[key] = globalFn[key] || {}
        globalFn[key][name || Object.keys(globalFn[key]).length] = fn
        parent.globalFn = globalFn
      }
    }
  
      
      const activateDarkMode = () => {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      const activateLightMode = () => {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }

      btf.activateDarkMode = activateDarkMode
      btf.activateLightMode = activateLightMode

      const theme = saveToLocal.get('theme')
    
          const mediaQueryDark = window.matchMedia('(prefers-color-scheme: dark)')
          const mediaQueryLight = window.matchMedia('(prefers-color-scheme: light)')
          
          if (theme === undefined) {
            if (mediaQueryLight.matches) activateLightMode()
            else if (mediaQueryDark.matches) activateDarkMode()
            else {
              const hour = new Date().getHours()
              const isNight = hour <= 6 || hour >= 18
              isNight ? activateDarkMode() : activateLightMode()
            }
            mediaQueryDark.addEventListener('change', () => {
              if (saveToLocal.get('theme') === undefined) {
                e.matches ? activateDarkMode() : activateLightMode()
              }
            })
          } else {
            theme === 'light' ? activateLightMode() : activateDarkMode()
          }
        
      
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        document.documentElement.classList.toggle('hide-aside', asideStatus === 'hide')
      }
    
      
    const detectApple = () => {
      if (/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)) {
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
  
    })()
  </script><script>const GLOBAL_CONFIG = {
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  translate: undefined,
  highlight: {"plugin":"highlight.js","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false,"highlightFullpage":false,"highlightMacStyle":false},
  copy: {
    success: 'Copy Successful',
    error: 'Copy Failed',
    noSupport: 'Browser Not Supported'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  dateSuffix: {
    just: 'Just now',
    min: 'minutes ago',
    hour: 'hours ago',
    day: 'days ago',
    month: 'months ago'
  },
  copyright: undefined,
  lightbox: 'medium_zoom',
  Snackbar: {"chs_to_cht":"You have switched to Traditional Chinese","cht_to_chs":"You have switched to Simplified Chinese","day_to_night":"You have switched to Dark Mode","night_to_day":"You have switched to Light Mode","bgLight":"#49b1f5","bgDark":"#2d3035","position":"top-center"},
  infinitegrid: {
    js: 'https://cdn.jsdelivr.net/npm/@egjs/infinitegrid/dist/infinitegrid.min.js',
    buttonText: 'Load More'
  },
  isPhotoFigcaption: true,
  islazyloadPlugin: false,
  isAnchor: false,
  percent: {
    toc: true,
    rightside: true,
  },
  autoDarkmode: true
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: '比较基因组',
  isHighlightShrink: false,
  isToc: true,
  pageType: 'post'
}</script><meta name="generator" content="Hexo 7.3.0"></head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img text-center"><img src="../../../../img/touxiang.jpg" onerror="this.onerror=null;this.src='../../../../img/friend_404.gif'" alt="avatar"/></div><div class="site-data text-center"><a href="../../../../archives/"><div class="headline">Articles</div><div class="length-num">3</div></a><a href="../../../../tags/"><div class="headline">Tags</div><div class="length-num">0</div></a><a href="../../../../categories/"><div class="headline">Categories</div><div class="length-num">0</div></a></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="../../../../index.html"><i class="fa-fw fas fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page" href="../../../../archives/"><i class="fa-fw fas fa-archive"></i><span> 归档</span></a></div><div class="menus_item"><a class="site-page" href="../../../../comments/"><i class="fa-fw fas fa-envelope-open"></i><span> 留言板</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url(/img/top_image.png);"><nav id="nav"><span id="blog-info"><a class="nav-site-title" href="../../../../index.html"><span class="site-name">VEXYXING</span></a><a class="nav-page-title" href="../../../../index.html"><span class="site-name">比较基因组</span></a></span><div id="menus"><div class="menus_items"><div class="menus_item"><a class="site-page" href="../../../../index.html"><i class="fa-fw fas fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page" href="../../../../archives/"><i class="fa-fw fas fa-archive"></i><span> 归档</span></a></div><div class="menus_item"><a class="site-page" href="../../../../comments/"><i class="fa-fw fas fa-envelope-open"></i><span> 留言板</span></a></div></div><div id="toggle-menu"><span class="site-page"><i class="fas fa-bars fa-fw"></i></span></div></div></nav><div id="post-info"><h1 class="post-title">比较基因组</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">Created</span><time class="post-meta-date-created" datetime="2025-02-12T06:16:46.000Z" title="Created 2025-02-12 14:16:46">2025-02-12</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">Updated</span><time class="post-meta-date-updated" datetime="2025-03-19T11:53:23.119Z" title="Updated 2025-03-19 19:53:23">2025-03-19</time></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title=""><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">Post Views:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="container post-content" id="article-container"><p>比较基因组各个步骤可能有多个方法都能达到相同目的，作为自学的，使用哪个方法是一个完全随缘的事情。</p>
<h1 id="1-系统发育分析"><a href="#1-系统发育分析" class="headerlink" title="1 系统发育分析"></a>1 系统发育分析</h1><h2 id="1-1-下载数据"><a href="#1-1-下载数据" class="headerlink" title="1.1 下载数据"></a>1.1 下载数据</h2><p>下载分析物种的<code>fna</code>基因组文件和<code>gff</code>注释文件。</p>
<p>过程如下：</p>
<ol>
<li><p>收集需下载的基因组登入号</p>
</li>
<li><p>组装下载链接</p>
</li>
<li><p>下载基因组和注释（需要安装ascp）</p>
</li>
</ol>
<p>由于不考虑自己重新注释的方法，我就将NCBI中某一个分类单位的所有可用的基因组全部下载下来（包括物种重复的）：</p>
<p><img src="/2025/02/12/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%842025-2-15/%E5%9B%BE1.png"></p>
<p>若在<code>Annotation列</code>有内容，则表示该基因组有可用的注释文件，就把<code>Genbank列</code>或<code>RefSeq列</code>复制到一个<code>id.txt</code>文档中，有<code>RefSeq</code>就优先复制<code>RefSeq</code>，没有就复制<code>GenBank</code>。如下图：</p>
<p><img src="/2025/02/12/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%842025-2-15/%E5%9B%BE2.png"></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">python assembly_genomic_download_link.py id.txt download.sh</span><br><span class="line"><span class="comment"># 运行组装脚本，如果下载失败的，会自动输出失败的登入号，可以单独拿出来重新组装。</span></span><br><span class="line"><span class="built_in">mkdir</span> workdir</span><br><span class="line"><span class="built_in">cd</span> workdir</span><br><span class="line">bash download.sh</span><br><span class="line">gunzip *.gz</span><br><span class="line"><span class="comment"># 开始下载，若某个基因组下载失败，可在次运行该指令，</span></span><br><span class="line"><span class="comment"># 已经下载好的基因组，它会自动跳过，没下载的基因组，他就会自动重新下载</span></span><br></pre></td></tr></table></figure>

<h2 id="1-2-筛选基因组"><a href="#1-2-筛选基因组" class="headerlink" title="1.2 筛选基因组"></a>1.2 筛选基因组</h2><p>一个基因可能有多个转录本，对后面<code>Othofinder</code>找同源单拷贝基因可能有影响，一般要去掉多余转录本，只留最长转录本。</p>
<p>此外，还应该把注释不好的基因组拿掉，并且，一个物种有多个基因组的时候，要筛选出注释最好的那个。</p>
<p>具体过程如下：</p>
<ol>
<li><p>先用extract_longest_transcript.py提取最长转录本的gff注释文件；</p>
</li>
<li><p>根据新的gff注释文件提所有编码基因的氨基酸序列；</p>
</li>
<li><p>利用rename_faa.py将氨基酸序列统一重命名</p>
</li>
<li><p>用BUSCO评估物种氨基酸序列的完整性</p>
</li>
<li><p>去除物种多余基因组、完整性较差基因组和最长转录本提取失败基因组</p>
</li>
</ol>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cd</span> ..</span><br><span class="line">python extract_longest_transcript.py ./workdir ./workdir</span><br><span class="line"><span class="built_in">cd</span> workdir</span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> *.fa</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">    file=<span class="variable">$&#123;i%.new.gff*&#125;</span></span><br><span class="line">    gffread <span class="string">&quot;<span class="variable">$&#123;file&#125;</span>.new.gff&#125;&quot;</span> -g <span class="string">&quot;<span class="variable">$&#123;file&#125;</span>.fna&quot;</span> -y <span class="string">&quot;<span class="variable">$&#123;file&#125;</span>.faa&quot;</span></span><br><span class="line">    python rename_faa.py  <span class="string">&quot;<span class="variable">$&#123;file&#125;</span>.faa&quot;</span> <span class="string">&quot;<span class="variable">$&#123;file&#125;</span>.rename.faa&quot;</span></span><br><span class="line">    busco -i <span class="string">&quot;<span class="variable">$&#123;file&#125;</span>.rename.faa&quot;</span> -m prot -c 40 -l ../busco_database/arthropoda_odb10 -o <span class="string">&quot;<span class="variable">$&#123;file&#125;</span>.busco.out&quot;</span> --offline</span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>
<h2 id="1-3-搜索单拷贝基因"><a href="#1-3-搜索单拷贝基因" class="headerlink" title="1.3 搜索单拷贝基因"></a>1.3 搜索单拷贝基因</h2><p>创建一个Orthofinder的工作目录，将所有合格的基因组的faa文件用cp命令复制到对应的文件夹中。然后使用Orthofinder搜索单拷贝基因。orthofinder搜索到的单拷贝基因存放在Single_Copy_Orthologue_Sequences文件夹中。</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">mkdir orthofinder</span><br><span class="line">cd orthofinder</span><br><span class="line">cp ../*.faa ./</span><br><span class="line"></span><br><span class="line">conda create -n orthofinder</span><br><span class="line">conda activate orthofinder</span><br><span class="line">conda install orthofinder -c bioconda</span><br><span class="line"></span><br><span class="line">orthofinder -f ./ -t 40</span><br></pre></td></tr></table></figure>
<h2 id="1-4-构建系统发生树"><a href="#1-4-构建系统发生树" class="headerlink" title="1.4 构建系统发生树"></a>1.4 构建系统发生树</h2><p>将Single_Copy_Orthologue_Sequences目录下的单拷贝基因使用mafft和Gblocks分别进行比对修剪，然后使用merge_align.py将单拷贝基因串联，最后使用Modeltest-ng和RAxML-ng建树。raxml-ng的输出结果利用iTOL网站在线进行查看和美化。</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">for i in *.fa</span><br><span class="line">do</span><br><span class="line">    file=$&#123;i%.fa*&#125;</span><br><span class="line">    mafft --maxiterate 1000 --localpair &quot;$&#123;file&#125;.fa&quot; &gt; &quot;$&#123;file&#125;.aln&quot;</span><br><span class="line">    Gblocks &quot;$&#123;file&#125;.aln&quot; -b4=5 -b5=h -t=p -e=.gb</span><br><span class="line">done</span><br><span class="line">python merge_align.py ./ ./</span><br><span class="line">modeltest-ng -i all.fa -d aa -o modeltest -p 40</span><br><span class="line">raxml-ng --msa all.fa --model TVM+G4 --threads 40 --prefix test --bs-trees 1000 --all</span><br></pre></td></tr></table></figure>
<h1 id="分歧时间估算"><a href="#分歧时间估算" class="headerlink" title="分歧时间估算"></a>分歧时间估算</h1><p>分歧时间目前文章都使用MCMCTREE，也有人先使用r8s，再使用MCMCtree。这里图r8s速度快，暂时先使用r8s。</p>
<h1 id="流程中使用的脚本"><a href="#流程中使用的脚本" class="headerlink" title="流程中使用的脚本"></a>流程中使用的脚本</h1><h2 id="assembly-genomic-download-link-py"><a href="#assembly-genomic-download-link-py" class="headerlink" title="assembly_genomic_download_link.py"></a>assembly_genomic_download_link.py</h2><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># assembly_genomic_download_link.py</span></span><br><span class="line"><span class="comment"># Usage:python assembly_genomic_download_link.py</span></span><br><span class="line"><span class="comment"># 使用之前，需要把``~/miniconda3/envs/download/etc/asperaweb_id_dsa.openssh``，换成你自己的位置。</span></span><br><span class="line"><span class="comment"># 这个是ascp的密钥，ascp可以更快下载ncbi上面的数据，比wget快非常多。没有安装的可以考虑安装一下。</span></span><br><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"><span class="keyword">import</span> requests</span><br><span class="line"><span class="keyword">import</span> re</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">read_txt</span>(<span class="params">path2txt</span>):</span><br><span class="line">    <span class="keyword">with</span> <span class="built_in">open</span>(path2txt, <span class="string">&#x27;r&#x27;</span>, encoding=<span class="string">&#x27;utf-8&#x27;</span>) <span class="keyword">as</span> file:</span><br><span class="line">        lines_uncleaned = file.readlines()</span><br><span class="line">    lines_uncleaned = [line.strip() <span class="keyword">for</span> line <span class="keyword">in</span> lines_uncleaned]</span><br><span class="line">    lines = []</span><br><span class="line">    <span class="keyword">for</span> line_2 <span class="keyword">in</span> lines_uncleaned:</span><br><span class="line">        <span class="keyword">if</span> <span class="string">&#x27;#&#x27;</span> <span class="keyword">not</span> <span class="keyword">in</span> line_2:</span><br><span class="line">            lines.append(line_2)</span><br><span class="line">    <span class="keyword">return</span> lines</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">url_prepare</span>(<span class="params">line</span>):</span><br><span class="line">    url_name = <span class="string">&#x27;https://www.ncbi.nlm.nih.gov/datasets/genome/&#x27;</span> + line</span><br><span class="line"></span><br><span class="line">    line_1 = line.split(<span class="string">&#x27;_&#x27;</span>)[<span class="number">0</span>]</span><br><span class="line">    line_2 = line.split(<span class="string">&#x27;_&#x27;</span>)[<span class="number">1</span>]</span><br><span class="line">    line_2_1 = line_2[:<span class="number">3</span>]</span><br><span class="line">    line_2_2 = line_2[<span class="number">3</span>:<span class="number">6</span>]</span><br><span class="line">    line_2_3 = line_2[<span class="number">6</span>:].split(<span class="string">&#x27;.&#x27;</span>)[<span class="number">0</span>]</span><br><span class="line">    url_fna = <span class="string">&#x27;https://ftp.ncbi.nlm.nih.gov/genomes/all/&#x27;</span> + line_1 + <span class="string">&quot;/&quot;</span> + line_2_1 + <span class="string">&quot;/&quot;</span> + line_2_2 + <span class="string">&quot;/&quot;</span> + line_2_3</span><br><span class="line"></span><br><span class="line">    <span class="keyword">return</span> url_name,url_fna</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">get_name</span>(<span class="params">url_name</span>):</span><br><span class="line">    response = requests.get(url_name)</span><br><span class="line">    html_content = response.text</span><br><span class="line">    match_name = <span class="built_in">str</span>(re.search(<span class="string">r&#x27;pageData\.speciesName = &quot;([^&quot;]+)&quot;&#x27;</span>, html_content)).split(<span class="string">&#x27;&quot;&#x27;</span>)[<span class="number">1</span>].replace(<span class="string">&#x27; &#x27;</span>,<span class="string">&#x27;_&#x27;</span>)</span><br><span class="line">    <span class="keyword">return</span> match_name</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">get_url</span>(<span class="params">url_fna,line</span>):</span><br><span class="line">    response = requests.get(url_fna)</span><br><span class="line">    html_content = response.text</span><br><span class="line">    ex1 = <span class="string">&#x27;&lt;a href=(.*?)/&lt;/a&gt;&#x27;</span></span><br><span class="line">    part_1 = <span class="built_in">str</span>(re.findall(ex1, html_content, re.S)).split(<span class="string">&#x27;&quot;&#x27;</span>)</span><br><span class="line">    aa = <span class="number">0</span></span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> part_1:</span><br><span class="line">        <span class="keyword">if</span> line <span class="keyword">in</span> i <span class="keyword">and</span> <span class="string">&#x27;]&#x27;</span> <span class="keyword">not</span> <span class="keyword">in</span> i <span class="keyword">and</span> aa == <span class="number">0</span>:</span><br><span class="line">            url_fna_2 = url_fna + <span class="string">&#x27;/&#x27;</span> + i + i.split(<span class="string">&#x27;/&#x27;</span>)[<span class="number">0</span>] + <span class="string">&#x27;_genomic.fna.gz&#x27;</span></span><br><span class="line">            aa += <span class="number">1</span></span><br><span class="line">            <span class="keyword">return</span> url_fna_2</span><br><span class="line">        <span class="keyword">elif</span> line <span class="keyword">in</span> i <span class="keyword">and</span> <span class="string">&#x27;]&#x27;</span> <span class="keyword">not</span> <span class="keyword">in</span> i <span class="keyword">and</span> aa != <span class="number">0</span>:</span><br><span class="line">            <span class="built_in">print</span>(<span class="string">&quot;error:&quot;</span>,line)</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">write_downloads_commend</span>(<span class="params">name,url,outpath,count,line</span>):</span><br><span class="line">    commend_fna = url.replace(<span class="string">&quot;https://ftp.ncbi.nlm.nih.gov&quot;</span>,</span><br><span class="line">     <span class="string">&quot;ascp -v -k 1 -T -l 200m -i ~/miniconda3/envs/download/etc/asperaweb_id_dsa.openssh anonftp@ftp.ncbi.nlm.nih.gov:&quot;</span>)\</span><br><span class="line">              + <span class="string">&#x27; ./&#x27;</span> + name + <span class="string">&#x27;_&#x27;</span> + line + <span class="string">&#x27;.fna.gz&#x27;</span></span><br><span class="line">    commend_gff = commend_fna.replace(<span class="string">&#x27;fna&#x27;</span>, <span class="string">&#x27;gff&#x27;</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">    <span class="keyword">if</span> count == <span class="number">0</span>:</span><br><span class="line">        <span class="keyword">with</span> <span class="built_in">open</span>(outpath, <span class="string">&quot;w&quot;</span>, encoding=<span class="string">&quot;utf-8&quot;</span>) <span class="keyword">as</span> file:</span><br><span class="line">            file.write(commend_fna.replace(<span class="string">&quot;&gt;&quot;</span>, <span class="string">&quot;&quot;</span>) + <span class="string">&quot;\n&quot;</span>)</span><br><span class="line">            file.write(commend_gff.replace(<span class="string">&quot;&gt;&quot;</span>, <span class="string">&quot;&quot;</span>) + <span class="string">&quot;\n&quot;</span>)</span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        <span class="keyword">with</span> <span class="built_in">open</span>(outpath, <span class="string">&quot;a&quot;</span>, encoding=<span class="string">&quot;utf-8&quot;</span>) <span class="keyword">as</span> file:</span><br><span class="line">            file.write(commend_fna.replace(<span class="string">&quot;&gt;&quot;</span>, <span class="string">&quot;&quot;</span>) + <span class="string">&quot;\n&quot;</span>)</span><br><span class="line">            file.write(commend_gff.replace(<span class="string">&quot;&gt;&quot;</span>, <span class="string">&quot;&quot;</span>) + <span class="string">&quot;\n&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> __name__ == <span class="string">&quot;__main__&quot;</span>:</span><br><span class="line">    path2txt = sys.argv[<span class="number">1</span>]</span><br><span class="line">    outpath = sys.argv[<span class="number">2</span>]</span><br><span class="line">    lines = read_txt(path2txt=path2txt)</span><br><span class="line">    number = <span class="built_in">len</span>(lines)</span><br><span class="line">    count = <span class="number">0</span></span><br><span class="line">    <span class="keyword">for</span> line <span class="keyword">in</span> lines:</span><br><span class="line">        <span class="keyword">try</span>:</span><br><span class="line"></span><br><span class="line">            url_name,url_fna = url_prepare(line)</span><br><span class="line">            name = get_name(url_name)</span><br><span class="line">            url = get_url(url_fna,line)</span><br><span class="line">            write_downloads_commend(name,url,outpath,count,line)</span><br><span class="line">            count += <span class="number">1</span></span><br><span class="line">            <span class="built_in">print</span>(<span class="built_in">str</span>(count) + <span class="string">&quot;/&quot;</span> + <span class="built_in">str</span>(number))</span><br><span class="line">        <span class="keyword">except</span> requests.RequestException <span class="keyword">as</span> e:</span><br><span class="line">            <span class="built_in">print</span>(line)</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">&quot;finished&quot;</span>)</span><br></pre></td></tr></table></figure>
<hr>
<h2 id="extract-longest-transcript-py"><a href="#extract-longest-transcript-py" class="headerlink" title="extract_longest_transcript.py"></a>extract_longest_transcript.py</h2><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># extract_longest_transcript.py</span></span><br><span class="line"><span class="comment"># Usage：python extract_longest_transcript.py &lt;input_folder&gt; &lt;output_folder&gt;</span></span><br><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> os</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">parse_gff</span>(<span class="params">file_path</span>):</span><br><span class="line">    data = []</span><br><span class="line">    <span class="keyword">with</span> <span class="built_in">open</span>(file_path, <span class="string">&#x27;r&#x27;</span>) <span class="keyword">as</span> f:</span><br><span class="line">        <span class="keyword">for</span> line <span class="keyword">in</span> f:</span><br><span class="line">            <span class="keyword">if</span> line.startswith(<span class="string">&quot;#&quot;</span>):</span><br><span class="line">                <span class="keyword">continue</span></span><br><span class="line">            parts = line.strip().split(<span class="string">&#x27;\t&#x27;</span>)</span><br><span class="line">            <span class="keyword">if</span> <span class="built_in">len</span>(parts) == <span class="number">9</span>:</span><br><span class="line">                attributes = &#123;key: value <span class="keyword">for</span> key, value <span class="keyword">in</span></span><br><span class="line">                              [attr.split(<span class="string">&#x27;=&#x27;</span>) <span class="keyword">for</span> attr <span class="keyword">in</span> parts[<span class="number">8</span>].split(<span class="string">&#x27;;&#x27;</span>) <span class="keyword">if</span> <span class="string">&#x27;=&#x27;</span> <span class="keyword">in</span> attr]&#125;</span><br><span class="line">                data.append(parts[:<span class="number">8</span>] + [attributes])</span><br><span class="line">    columns = [<span class="string">&#x27;seqid&#x27;</span>, <span class="string">&#x27;source&#x27;</span>, <span class="string">&#x27;type&#x27;</span>, <span class="string">&#x27;start&#x27;</span>, <span class="string">&#x27;end&#x27;</span>, <span class="string">&#x27;score&#x27;</span>, <span class="string">&#x27;strand&#x27;</span>, <span class="string">&#x27;phase&#x27;</span>, <span class="string">&#x27;attributes&#x27;</span>]</span><br><span class="line">    <span class="keyword">return</span> pd.DataFrame(data, columns=columns)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">select_longest_transcript</span>(<span class="params">df</span>):</span><br><span class="line">    transcripts = df[df[<span class="string">&#x27;type&#x27;</span>] == <span class="string">&#x27;mRNA&#x27;</span>].copy()</span><br><span class="line">    transcripts[<span class="string">&#x27;length&#x27;</span>] = transcripts.apply(<span class="keyword">lambda</span> row: <span class="built_in">int</span>(row[<span class="string">&#x27;end&#x27;</span>]) - <span class="built_in">int</span>(row[<span class="string">&#x27;start&#x27;</span>]), axis=<span class="number">1</span>)</span><br><span class="line"></span><br><span class="line">    transcripts[<span class="string">&#x27;transcript_id&#x27;</span>] = transcripts[<span class="string">&#x27;attributes&#x27;</span>].apply(<span class="keyword">lambda</span> x: x.get(<span class="string">&#x27;ID&#x27;</span>))</span><br><span class="line">    transcripts[<span class="string">&#x27;gene_id&#x27;</span>] = transcripts[<span class="string">&#x27;attributes&#x27;</span>].apply(<span class="keyword">lambda</span> x: x.get(<span class="string">&#x27;Parent&#x27;</span>))</span><br><span class="line"></span><br><span class="line">    longest_transcripts = transcripts.loc[transcripts.groupby(<span class="string">&#x27;gene_id&#x27;</span>)[<span class="string">&#x27;length&#x27;</span>].idxmax()]</span><br><span class="line"></span><br><span class="line">    longest_ids = <span class="built_in">set</span>(longest_transcripts[<span class="string">&#x27;transcript_id&#x27;</span>])</span><br><span class="line">    gene_ids = <span class="built_in">set</span>(df[df[<span class="string">&#x27;type&#x27;</span>] == <span class="string">&#x27;gene&#x27;</span>][<span class="string">&#x27;attributes&#x27;</span>].apply(<span class="keyword">lambda</span> x: x.get(<span class="string">&#x27;ID&#x27;</span>)))</span><br><span class="line">    region_ids = <span class="built_in">set</span>(df[df[<span class="string">&#x27;type&#x27;</span>] == <span class="string">&#x27;region&#x27;</span>][<span class="string">&#x27;attributes&#x27;</span>].apply(<span class="keyword">lambda</span> x: x.get(<span class="string">&#x27;ID&#x27;</span>)))</span><br><span class="line"></span><br><span class="line">    longest_df = df[df[<span class="string">&#x27;attributes&#x27;</span>].apply(<span class="keyword">lambda</span> x: x.get(<span class="string">&#x27;Parent&#x27;</span>) <span class="keyword">in</span> longest_ids</span><br><span class="line">                                                     <span class="keyword">or</span> x.get(<span class="string">&#x27;ID&#x27;</span>) <span class="keyword">in</span> longest_ids</span><br><span class="line">                                                     <span class="keyword">or</span> x.get(<span class="string">&#x27;ID&#x27;</span>) <span class="keyword">in</span> gene_ids</span><br><span class="line">                                                     <span class="keyword">or</span> x.get(<span class="string">&#x27;ID&#x27;</span>) <span class="keyword">in</span> region_ids)]</span><br><span class="line">    <span class="keyword">return</span> longest_df</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">write_gff</span>(<span class="params">df, output_file</span>):</span><br><span class="line">    <span class="keyword">with</span> <span class="built_in">open</span>(output_file, <span class="string">&#x27;w&#x27;</span>) <span class="keyword">as</span> f:</span><br><span class="line">        <span class="keyword">for</span> _, row <span class="keyword">in</span> df.iterrows():</span><br><span class="line">            attributes = <span class="string">&#x27;;&#x27;</span>.join([<span class="string">f&quot;<span class="subst">&#123;key&#125;</span>=<span class="subst">&#123;value&#125;</span>&quot;</span> <span class="keyword">for</span> key, value <span class="keyword">in</span> row[<span class="string">&#x27;attributes&#x27;</span>].items()])</span><br><span class="line">            line = <span class="string">&#x27;\t&#x27;</span>.join(<span class="built_in">map</span>(<span class="built_in">str</span>, row[:<span class="number">8</span>])) + <span class="string">&#x27;\t&#x27;</span> + attributes + <span class="string">&#x27;\n&#x27;</span></span><br><span class="line">            f.write(line)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">input_folder = sys.argv[<span class="number">1</span>]</span><br><span class="line">output_folder = sys.argv[<span class="number">2</span>]</span><br><span class="line">os.makedirs(output_folder, exist_ok=<span class="literal">True</span>)</span><br><span class="line"><span class="keyword">for</span> file_name <span class="keyword">in</span> os.listdir(input_folder):</span><br><span class="line">    <span class="keyword">if</span> file_name.endswith(<span class="string">&quot;.gff&quot;</span>):</span><br><span class="line">        input_gff = os.path.join(input_folder, file_name)</span><br><span class="line">        output_gff = os.path.join(output_folder, file_name.replace(<span class="string">&quot;.gff&quot;</span>, <span class="string">&quot;.new.gff&quot;</span>))</span><br><span class="line"></span><br><span class="line">    gff_df = parse_gff(input_gff)</span><br><span class="line">    longest_df = select_longest_transcript(gff_df)</span><br><span class="line">    write_gff(longest_df, output_gff)</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<hr>
<h2 id="rename-faa-py"><a href="#rename-faa-py" class="headerlink" title="rename_faa.py"></a>rename_faa.py</h2><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># rename_faa.py</span></span><br><span class="line"><span class="comment">#Usage:python rename_faa.py &lt;input_file&gt; &lt;output_file&gt;</span></span><br><span class="line"><span class="keyword">import</span> os</span><br><span class="line"><span class="keyword">from</span> Bio <span class="keyword">import</span> SeqIO</span><br><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">rename_sequences_by_filename</span>(<span class="params">input_file, output_file</span>):</span><br><span class="line">    base_name = os.path.basename(input_file).split(<span class="string">&quot;.&quot;</span>)[<span class="number">0</span>]</span><br><span class="line">    </span><br><span class="line">    sequences = <span class="built_in">list</span>(SeqIO.parse(input_file, <span class="string">&quot;fasta&quot;</span>))</span><br><span class="line"></span><br><span class="line">    <span class="keyword">for</span> i, record <span class="keyword">in</span> <span class="built_in">enumerate</span>(sequences, start=<span class="number">1</span>):</span><br><span class="line">        record.<span class="built_in">id</span> = <span class="string">f&quot;<span class="subst">&#123;base_name&#125;</span>_<span class="subst">&#123;i:05d&#125;</span>&quot;</span></span><br><span class="line">        record.description = <span class="string">&quot;&quot;</span>  </span><br><span class="line">        record.seq = record.seq.replace(<span class="string">&quot;.&quot;</span>, <span class="string">&quot;-&quot;</span>)</span><br><span class="line"></span><br><span class="line">    SeqIO.write(sequences, output_file, <span class="string">&quot;fasta&quot;</span>)</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">f&quot;Processed file: <span class="subst">&#123;file_name&#125;</span>, renamed <span class="subst">&#123;<span class="built_in">len</span>(sequences)&#125;</span> sequences.&quot;</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment"># 示例用法</span></span><br><span class="line"><span class="keyword">if</span> __name__ == <span class="string">&quot;__main__&quot;</span>:</span><br><span class="line">    input_file = sys.argv[<span class="number">1</span>]  <span class="comment"># 替换为包含 .faa 文件的文件夹路径</span></span><br><span class="line">    output_file = sys.argv[<span class="number">1</span>]  <span class="comment"># 替换为保存结果的文件夹路径</span></span><br><span class="line">    rename_sequences_by_filename(input_file, output_file)</span><br></pre></td></tr></table></figure>
<hr>
<h2 id="merge-align-py"><a href="#merge-align-py" class="headerlink" title="merge_align.py"></a>merge_align.py</h2><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># merge_align.py</span></span><br><span class="line"><span class="comment"># Usage:python merge_align.py &lt;input_dir&gt; &lt;output_dir&gt;</span></span><br><span class="line"><span class="keyword">import</span> os</span><br><span class="line"><span class="keyword">from</span> Bio <span class="keyword">import</span> SeqIO</span><br><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">merge_genome</span>(<span class="params">inputfile_path, outputfile_path</span>):</span><br><span class="line"></span><br><span class="line">    merge_records = []</span><br><span class="line">    count = <span class="number">0</span></span><br><span class="line"></span><br><span class="line">    <span class="keyword">for</span> filename <span class="keyword">in</span> os.listdir(inputfile_path):</span><br><span class="line">        <span class="keyword">if</span> filename.endswith(<span class="string">&quot;.gb&quot;</span>):</span><br><span class="line">            input_file = os.path.join(inputfile_path, filename)</span><br><span class="line">            </span><br><span class="line">            records = SeqIO.parse(input_file, <span class="string">&quot;fasta&quot;</span>)</span><br><span class="line">            <span class="keyword">if</span> count == <span class="number">0</span>:</span><br><span class="line">                <span class="keyword">for</span> record <span class="keyword">in</span> records:</span><br><span class="line">                    record.<span class="built_in">id</span> = record.<span class="built_in">id</span>.split(<span class="string">&quot;_genomic&quot;</span>)[<span class="number">0</span>]</span><br><span class="line">                    merge_records.append(record)</span><br><span class="line">                count += <span class="number">1</span></span><br><span class="line">            <span class="keyword">else</span>:</span><br><span class="line">                merge_records_tmp = []</span><br><span class="line">                <span class="keyword">for</span> record <span class="keyword">in</span> records:</span><br><span class="line">                    <span class="keyword">for</span> old_record <span class="keyword">in</span> merge_records:</span><br><span class="line">                        <span class="keyword">if</span>  old_record.<span class="built_in">id</span> == record.<span class="built_in">id</span>.split(<span class="string">&quot;_genomic&quot;</span>)[<span class="number">0</span>]:</span><br><span class="line">                            new_record = old_record</span><br><span class="line">                            new_record.seq = old_record.seq + record.seq</span><br><span class="line">                            merge_records_tmp.append(new_record)</span><br><span class="line">                        </span><br><span class="line"></span><br><span class="line">                            </span><br><span class="line">                count += <span class="number">1</span></span><br><span class="line">                merge_records = merge_records_tmp</span><br><span class="line">        <span class="keyword">break</span></span><br><span class="line">    output_file = os.path.join(outputfile_path, <span class="string">&quot;all.fa&quot;</span>)</span><br><span class="line">    SeqIO.write(merge_records, output_file, <span class="string">&quot;fasta&quot;</span>)</span><br><span class="line">    <span class="built_in">print</span>(count)</span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> __name__ == <span class="string">&#x27;__main__&#x27;</span>:</span><br><span class="line"></span><br><span class="line">    inputfile_path = sys.argv[<span class="number">1</span>]</span><br><span class="line">    outputfile_path = sys.argv[<span class="number">2</span>]</span><br><span class="line">    merge_genome(inputfile_path, outputfile_path)</span><br><span class="line"></span><br></pre></td></tr></table></figure></article><div class="tag_share"><div class="post-share"><div class="social-share" data-image="../../../../img/touxiang.jpg" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><a class="pagination-related full-width" href="../../../03/18/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2%E5%88%86%E6%9E%90/" title="系统发育分析.md"><div class="cover" style="background: var(--default-bg-color)"></div><div class="info text-right"><div class="info-1"><div class="info-item-1">Next</div><div class="info-item-2">系统发育分析.md</div></div><div class="info-2"><div class="info-item-1">概述为了寻求各物种之间的系统发育关系，而去寻找各物种之间共有的同源单拷贝基因。我认为难点在于筛选数据，目前指定筛选的顺序为：  基因组：在一个目标分类单位中，如螯肢动物亚门，寻找其中所有可用的基因组数据，使用compleasm进行检验，保证使用的基因组具备高完整性和低冗余性，从而有较多的同源单拷贝基因可供选择。这里的阈值（S）我选择90%。即为第一轮筛选。  此时，得到一批基因组数据及其一组同源单拷贝基因，同时，提取一个物种的这组同源单拷贝基因作为参考。  基因组二代测序数据：这是第二轮数据筛选，从NCBI上下载缺失数据的分类单位的二代测序数据，然后使用参考的单拷贝基因数据进行搜索该物种对应的单拷贝基因数据，软件使用的是aTRAM。  转录组测序数据：这是第三轮数据筛选，即使用Trinity组装转录本，提取单拷贝基因   1 一轮筛选：基因组1.1...</div></div></div></a></nav></div><div class="aside-content" id="aside-content"><div class="card-widget card-info text-center"><div class="avatar-img"><img src="../../../../img/touxiang.jpg" onerror="this.onerror=null;this.src='../../../../img/friend_404.gif'" alt="avatar"/></div><div class="author-info-name">VERYXING565</div><div class="author-info-description">有事邮箱：veryxing565@163.com</div><div class="site-data"><a href="../../../../archives/"><div class="headline">Articles</div><div class="length-num">3</div></a><a href="../../../../tags/"><div class="headline">Tags</div><div class="length-num">0</div></a><a href="../../../../categories/"><div class="headline">Categories</div><div class="length-num">0</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/veryxing565"><i class="fab fa-github"></i><span>Follow Me</span></a></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>Contents</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#1-%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2%E5%88%86%E6%9E%90"><span class="toc-number">1.</span> <span class="toc-text">1 系统发育分析</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#1-1-%E4%B8%8B%E8%BD%BD%E6%95%B0%E6%8D%AE"><span class="toc-number">1.1.</span> <span class="toc-text">1.1 下载数据</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#1-2-%E7%AD%9B%E9%80%89%E5%9F%BA%E5%9B%A0%E7%BB%84"><span class="toc-number">1.2.</span> <span class="toc-text">1.2 筛选基因组</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#1-3-%E6%90%9C%E7%B4%A2%E5%8D%95%E6%8B%B7%E8%B4%9D%E5%9F%BA%E5%9B%A0"><span class="toc-number">1.3.</span> <span class="toc-text">1.3 搜索单拷贝基因</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#1-4-%E6%9E%84%E5%BB%BA%E7%B3%BB%E7%BB%9F%E5%8F%91%E7%94%9F%E6%A0%91"><span class="toc-number">1.4.</span> <span class="toc-text">1.4 构建系统发生树</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%88%86%E6%AD%A7%E6%97%B6%E9%97%B4%E4%BC%B0%E7%AE%97"><span class="toc-number">2.</span> <span class="toc-text">分歧时间估算</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%B5%81%E7%A8%8B%E4%B8%AD%E4%BD%BF%E7%94%A8%E7%9A%84%E8%84%9A%E6%9C%AC"><span class="toc-number">3.</span> <span class="toc-text">流程中使用的脚本</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#assembly-genomic-download-link-py"><span class="toc-number">3.1.</span> <span class="toc-text">assembly_genomic_download_link.py</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#extract-longest-transcript-py"><span class="toc-number">3.2.</span> <span class="toc-text">extract_longest_transcript.py</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#rename-faa-py"><span class="toc-number">3.3.</span> <span class="toc-text">rename_faa.py</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#merge-align-py"><span class="toc-number">3.4.</span> <span class="toc-text">merge_align.py</span></a></li></ol></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>Recent Posts</span></div><div class="aside-list"><div class="aside-list-item no-cover"><div class="content"><a class="title" href="../../../05/05/hybpiper%E4%B8%AA%E4%BA%BA%E4%BD%BF%E7%94%A8%E8%AF%B4%E6%98%8E/" title="hypiper个人使用说明">hypiper个人使用说明</a><time datetime="2025-05-05T06:16:46.000Z" title="Created 2025-05-05 14:16:46">2025-05-05</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="../../../03/18/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2%E5%88%86%E6%9E%90/" title="系统发育分析.md">系统发育分析.md</a><time datetime="2025-03-18T04:24:34.000Z" title="Created 2025-03-18 12:24:34">2025-03-18</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="" title="比较基因组">比较基因组</a><time datetime="2025-02-12T06:16:46.000Z" title="Created 2025-02-12 14:16:46">2025-02-12</time></div></div></div></div></div></div></main><footer id="footer" style="background-image: url(/img/footer.jpg);"><div id="footer-wrap"><div class="copyright">&copy;2025 By VERYXING565</div><div class="framework-info"><span>Framework </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>Theme </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div><div class="footer_custom_text">Hi, welcome to my <a target="_blank" rel="noopener" href="https://butterfly.js.org/">blog</a>!</div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="Reading Mode"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="Toggle Between Light and Dark Mode"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="Toggle Between Single-column and Double-column"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside-config" type="button" title="Settings"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="Table of Contents"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="Back to Top"><span class="scroll-percent"></span><i class="fas fa-arrow-up"></i></button></div></div><div><script src="../../../../js/utils.js"></script><script src="../../../../js/main.js"></script><script src="https://cdn.jsdelivr.net/npm/medium-zoom/dist/medium-zoom.min.js"></script><script src="https://cdn.jsdelivr.net/npm/node-snackbar/dist/snackbar.min.js"></script><div class="js-pjax"></div><canvas class="fireworks" mobile="false"></canvas><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/dist/fireworks.min.js"></script><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>